source/docs/v3/tenancy-bootstrappers.blade.md

```diff
@@ -29,31 +29,77 @@ php artisan cache:clear --tag=tenant_123
 
-    public function boot()
-    {
-        // update the middleware used by the asset controller
-        TenantAssetsController::$tenancyMiddleware = InitializeTenancyByDomainOrSubdomain::class;
-    }
+### Assets
+
+The bootstrapper modifies the links retrieved by the `asset()` helper, so they link to the files *of the currently initialized and identified tenant*.
+
+Before using the `asset()` helper, make sure to [assign the identification middleware you're using in your app to TenantAssetsController's `$tenancyMiddleware`]({{ $page->link('configuration#static-properties') }}):
+
+```php
+// TenancyServiceProvider (don't forget to import the classes)
+
+public function boot()
+{
+    // Update the middleware used by the asset controller
+    TenantAssetsController::$tenancyMiddleware = InitializeTenancyByDomainOrSubdomain::class;
+}
 ```
 
-This bootstrapper is the most complex one, by far. We will have a — better written — explanation in v3 docs soon, but for now, refer to the 2.x docs for information about filesystem tenancy. [https://tenancyforlaravel.com/docs/v2/filesystem-tenancy/](https://tenancyforlaravel.com/docs/v2/filesystem-tenancy/)
+> Note: You can disable tenancy of `asset()` in the config (`tenancy.filesystem.asset_helper_tenancy`) and explicitly use `tenant_asset()` instead. But keep in mind that `tenant_asset()` does not respect the asset URL you set in `config('app.asset_url')` – the helper will always return your website's url suffixed by `tenancy/assets/name-of-your-asset.foo` (e.g. `https://your-site.test/tenancy/assets/asset-name.txt`).
+>
+> You may want to do that if you're facing issues using a package that utilizes `asset()` inside the tenant app.
+>
+> For non-tenant-specific assets (assets shared among all tenants or JS/CSS assets), you can use `global_asset()` and `mix()`.
+
+If `config('app.asset_url')` has been set, the bootstrapper suffixes the configured asset URL the same way as `storage_path()` (useful if you're using Laravel Vapor or similar).
 
-If you don't want to bootstrap filesystem tenancy in this way, and want to — for example — provision an S3 bucket for each tenant, you can absolutely do that. Look at the package's bootstrappers to get an idea of how to write one yourself, and feel free to implement it any way you want.
+If `config('app.asset_url')` is null (as it is by default), `asset()` will return a URL pointing to the TenantAssetsController (`/tenancy/assets/...`) which returns a file response:
+
+```php
+// TenantAssetsController
+return response()->file(storage_path('app/public/' . $path));
+```
+
+When `config('app.asset_url')` is null, **you need to store the assets in your tenant's `app/public/` directory**. So for non-private (global) assets, you may want to create a disk and use URLs from that disk instead (don't add that disk to `config(tenancy.filesystem.disks)`). For example:
```

Comment:
This section is still a bit messy:
1. This paragraph directly follows another paragraph that starts basically the same: `If `config('app.asset_url')` is null`. It *feels* like it's a typo because the flow of the page suggests that this paragraph would be put in contrast against the previous paragraph, but they're talking about the same thing. When you read the section for the second time, it does make sense, but on first read it seems confusing.
2. There's a lot of text explaining edge cases or configuration *before* the actual behavior is explained.

I'd restructure this entire section. Start with what the package does out of the box, then talk about what happens when `ASSET_URL` is **not** null, then talk about global assets, then mention disabling asset helper tenancy, and then explain using the `tenant_asset()` helper.

The last few sections can be swapped depending on what you feel is best, but I'd definitely start with explaining what is being done out of the box, then explaining the **common** complexities that people may have to deal with (global assets, `ASSET_URL` being set by Vapor), and only then talk about the stuff you have in the note at the top. Maybe `ASSET_URL` can be at the very end since that's a Vapor-specific complexity, and much less relevant than something like global assets which pretty much everyone has to deal with.

I'd also first explain what the helper does, and only then mention the `TenantAssetsController::$tenancyMiddleware`. We first want to tell people **what this section is about**, and only then talk about all of the nuances and complexities of it.

Reply:
Thanks for the guidance, I restructured the section https://github.com/stancl/tenancy-docs/pull/201/commits/3263e2bb50a2d63028ea431aef88c7c0a00170b7